monkey/tests/unit_tests/monkey_island/cc/island_event_handlers/test_handle_agent_registration.py

```diff
@@ -0,0 +1,235 @@
+
+    with pytest.raises(Exception):
+        handler(AGENT_REGISTRATION_DATA)
+
+
+def test_add_agent(handler, agent_repository):
+    expected_agent = Agent(
+        id=AGENT_REGISTRATION_DATA.id,
+        machine_id=SEED_ID,
+        start_time=AGENT_REGISTRATION_DATA.start_time,
+        parent_id=AGENT_REGISTRATION_DATA.parent_id,
+        cc_server=AGENT_REGISTRATION_DATA.cc_server,
+    )
+    handler(AGENT_REGISTRATION_DATA)
+
+    agent_repository.upsert_agent.assert_called_with(expected_agent)
+
+
+def test_add_node_connection(handler, machine_repository, node_repository):
+    island_machine = Machine(
+        id=1,
+        hardware_id=99,
+        island=True,
+        network_interfaces=[IPv4Interface("192.168.1.1/24")],
+    )
+    get_machines_by_ip = build_get_machines_by_ip(
+        island_machine.network_interfaces[0].ip, island_machine
+    )
+    machine_repository.get_machines_by_ip = MagicMock(side_effect=get_machines_by_ip)
+    machine_repository.get_machine_by_hardware_id = MagicMock(return_value=MACHINE)
+
+    handler(AGENT_REGISTRATION_DATA)
+
+    node_repository.upsert_communication.assert_called_once()
+    node_repository.upsert_communication.assert_called_with(
+        MACHINE.id, island_machine.id, CommunicationType.CC
+    )
+
+
+def test_add_node_connection__unknown_server(handler, machine_repository, node_repository):
+    expected_new_server_machine = Machine(
+        id=SEED_ID,
+        network_interfaces=[IPv4Interface("192.168.1.1/32")],
+    )
+
+    machine_repository.get_machine_by_hardware_id = MagicMock(return_value=MACHINE)
+    handler(AGENT_REGISTRATION_DATA)
+
+    machine_repository.upsert_machine.assert_called_with(expected_new_server_machine)
+    node_repository.upsert_communication.assert_called_with(
+        MACHINE.id, SEED_ID, CommunicationType.CC
+    )
+
+
+def test_machine_interfaces_updated(handler, machine_repository):
+    existing_machine = Machine(
+        id=SEED_ID,
+        hardware_id=AGENT_REGISTRATION_DATA.machine_hardware_id,
+        network_interfaces=[IPv4Interface("192.168.1.2/32"), IPv4Interface("192.168.1.5/32")],
+    )
+    machine_repository.get_machine_by_hardware_id = MagicMock(return_value=existing_machine)
+    agent_registration_data = AgentRegistrationData(
+        id=AGENT_ID,
+        machine_hardware_id=MACHINE.hardware_id,
+        start_time=0,
+        parent_id=None,
+        cc_server="192.168.1.1:5000",
+        network_interfaces=[
+            IPv4Interface("192.168.1.2/24"),
+            IPv4Interface("192.168.1.3/16"),
+            IPv4Interface("192.168.1.4/24"),
+        ],
+    )
+    expected_network_interfaces = tuple(
+        sorted(
+            (*agent_registration_data.network_interfaces, existing_machine.network_interfaces[-1])
+        )
+    )
+
+    handler(agent_registration_data)
+    updated_machine = machine_repository.upsert_machine.call_args_list[0][0][0]
```

Comment:
Is there a way to do this that is more readable than `call_args_list[0][0][0]`?

Reply:
Not that I can find. Every time I have to deal with the `call_args_list` I have to remember how many levels of garbage you have to wade down through.

I mean, I can do 
```
call = [0]
args = call[0]
updated_machine = args[0]
```